File: frontend/src/lib/api/autonomy.ts
import { fetchJson } from './client';

export type RuleType = 'global' | 'domain' | 'contact' | 'tag';
export type AutonomyDecision = 'auto_apply' | 'defer' | 'block' | 'queue_for_later';

export interface AutonomyRule {
	id: string;
	rule_type: RuleType;
	scope_key: string;
	auto_apply_threshold: number;
	max_actions_per_hour: number;
	allowed_intent_types: string[];
	blocked_intent_types: string[];
	quiet_hours_start: string | null;
	quiet_hours_end: string | null;
	quiet_hours_timezone: string | null;
	enabled: boolean;
	created_at: string;
}

export interface ActionLogEntry {
	id: string;
	rule_id: string | null;
	intent_type: string;
	decision: AutonomyDecision;
	confidence: number;
	reason: string;
	created_at: string;
}

export async function listRules(): Promise<AutonomyRule[]> {
	const res = await fetchJson<{ rules: AutonomyRule[] }>('/api/v1/autonomy/rules');
	return res.rules;
}

export async function createRule(data: Partial<AutonomyRule>): Promise<AutonomyRule> {
	return fetchJson<AutonomyRule>('/api/v1/autonomy/rules', {
		method: 'POST',
		body: JSON.stringify(data)
	});
}

export async function getRule(id: string): Promise<AutonomyRule> {
	return fetchJson<AutonomyRule>(`/api/v1/autonomy/rules/${id}`);
}

export async function updateRule(id: string, data: Partial<AutonomyRule>): Promise<AutonomyRule> {
	return fetchJson<AutonomyRule>(`/api/v1/autonomy/rules/${id}`, {
		method: 'PUT',
		body: JSON.stringify(data)
	});
}

export async function deleteRule(id: string): Promise<void> {
	await fetchJson<void>(`/api/v1/autonomy/rules/${id}`, { method: 'DELETE' });
}

export async function listActionLog(limit = 50, offset = 0): Promise<ActionLogEntry[]> {
	const res = await fetchJson<{ entries: ActionLogEntry[] }>(
		`/api/v1/autonomy/action-log?limit=${limit}&offset=${offset}`
	);
	return res.entries;
}

File: crates/mv-engine/src/autonomy.rs
use std::sync::Arc;

use chrono::{NaiveTime, Utc};
use mv_core::*;
use mv_storage::unified::UnifiedStore;
use uuid::Uuid;

pub struct AutonomyGate {
    store: Arc<UnifiedStore>,
}

impl AutonomyGate {
    pub fn new(store: Arc<UnifiedStore>) -> Self {
        Self { store }
    }

    /// Evaluate an action against autonomy rules.
    /// Rules cascade: contact > domain > tag > global.
    pub async fn evaluate(
        &self,
        intent_type: &str,
        confidence: f32,
        scope_hints: &[(&str, &str)], // [(rule_type, scope_key)]
    ) -> MvResult<AutonomyDecision> {
        let rules = self.store.nodes.list_autonomy_rules().await?;
        let now = Utc::now();

        // Priority order: find most specific matching rule
        let priority_order = ["contact", "domain", "tag", "global"];

        for rule_type in priority_order {
            let matching_rules: Vec<&AutonomyRule> = rules
                .iter()
                .filter(|r| r.enabled && r.rule_type == rule_type)
                .filter(|r| {
                    if rule_type == "global" {
                        true
                    } else {
                        scope_hints
                            .iter()
                            .any(|(rt, sk)| *rt == rule_type && r.scope_key.as_deref() == Some(*sk))
                    }
                })
                .collect();

            if let Some(rule) = matching_rules.first() {
                return self.apply_rule(rule, intent_type, confidence, now).await;
            }
        }

        // No rules match -- defer by default
        Ok(AutonomyDecision::Defer)
    }

    async fn apply_rule(
        &self,
        rule: &AutonomyRule,
        intent_type: &str,
        confidence: f32,
        now: chrono::DateTime<Utc>,
    ) -> MvResult<AutonomyDecision> {
        // Check blocked intent types
        if !rule.blocked_intent_types.is_empty()
            && rule.blocked_intent_types.iter().any(|t| t == intent_type)
        {
            self.log_action(
                Some(rule.id),
                intent_type,
                AutonomyDecision::Block,
                confidence,
                "blocked intent type",
            )
            .await?;
            return Ok(AutonomyDecision::Block);
        }

        // Check allowed intent types (if specified, must match)
        if !rule.allowed_intent_types.is_empty()
            && !rule.allowed_intent_types.iter().any(|t| t == intent_type)
        {
            self.log_action(
                Some(rule.id),
                intent_type,
                AutonomyDecision::Defer,
                confidence,
                "intent type not in allowed list",
            )
            .await?;
            return Ok(AutonomyDecision::Defer);
        }

        // Check quiet hours
        if self.in_quiet_hours(rule, now) {
            self.log_action(
                Some(rule.id),
                intent_type,
                AutonomyDecision::QueueForLater,
                confidence,
                "quiet hours",
            )
            .await?;
            return Ok(AutonomyDecision::QueueForLater);
        }

        // Check rate limit
        let one_hour_ago = now - chrono::Duration::hours(1);
        let recent_count = self
            .store
            .nodes
            .count_recent_actions(Some(rule.id), one_hour_ago)
            .await?;
        if recent_count >= rule.max_actions_per_hour as usize {
            self.log_action(
                Some(rule.id),
                intent_type,
                AutonomyDecision::Defer,
                confidence,
                "rate limit exceeded",
            )
            .await?;
            return Ok(AutonomyDecision::Defer);
        }

        // Check confidence threshold
        if confidence >= rule.auto_apply_threshold {
            self.log_action(
                Some(rule.id),
                intent_type,
                AutonomyDecision::AutoApply,
                confidence,
                "above threshold",
            )
            .await?;
            return Ok(AutonomyDecision::AutoApply);
        }

        self.log_action(
            Some(rule.id),
            intent_type,
            AutonomyDecision::Defer,
            confidence,
            "below threshold",
        )
        .await?;
        Ok(AutonomyDecision::Defer)
    }

    fn in_quiet_hours(&self, rule: &AutonomyRule, now: chrono::DateTime<Utc>) -> bool {
        let (start_str, end_str) = match (&rule.quiet_hours_start, &rule.quiet_hours_end) {
            (Some(s), Some(e)) => (s.as_str(), e.as_str()),
            _ => return false,
        };

        let start = match NaiveTime::parse_from_str(start_str, "%H:%M") {
            Ok(t) => t,
            Err(_) => return false,
        };
        let end = match NaiveTime::parse_from_str(end_str, "%H:%M") {
            Ok(t) => t,
            Err(_) => return false,
        };

        // Use UTC time (timezone handling is simplified)
        let current = now.time();

        if start <= end {
            current >= start && current < end
        } else {
            // Wraps midnight
            current >= start || current < end
        }
    }

    async fn log_action(
        &self,
        rule_id: Option<Uuid>,
        intent_type: &str,
        decision: AutonomyDecision,
        confidence: f32,
        reason: &str,
    ) -> MvResult<()> {
        let log = AutonomyActionLog {
            id: Uuid::new_v4(),
            rule_id,
            intent_type: intent_type.to_string(),
            decision,
            confidence: Some(confidence),
            reason: Some(reason.to_string()),
            created_at: Utc::now(),
        };
        self.store.nodes.log_autonomy_action(&log).await
    }
}
